### PR TITLE
Abstract System Level Actions for iOS into a Action Library

### PR DESCRIPTION
**Pull request by lilyzh**
Abstract System Level Actions for iOS into a Action Library

JIRA: 
https://mobify.atlassian.net/browse/CSOG-4

## Changes
Added iosAction.js file under pageObjects folder, including:

- Generalize the native xpath selectors
- Generalize native action functions
- Navigate between tabs
- Use the system level back button
- Generalize setting the context object

## How to test-drive this PR
- Clone https://github.com/mobify/generator-astro
- Change `generator.sh`:
```
SCAFFOLD_VERSION="CSOG-4_system_action_ios"
SCAFFOLD_URL="https://github.com/mobify/astro-scaffold/archive/CSOG-4_system_action_ios.zip"
```
- Run `./generator.sh`
- `cd` into generated project and run `./scripts/ios-appium.sh`

## TODOS:
- [x] Tests are passing on your machine.
- [x] Change works in both Android and iOS. (Tested with iOS build and test script. Not sure if there's an equivalent for Android). 
- [x] CHANGELOG.md has been updated.
- [x] Scaffold is working. If a new feature was added, it was added to the Scaffold app.
- [x] Run the generator to make sure it still functions correctly.
- [x] QA +1
- [ ] +1 from an engineer on the Astro team. @jeremywiebe

### CODE REVIEW

**Comment by ellenmobify** [comment]
How do you feel about renaming `iosAction` to `iosActions`?

File: app/tests/system/pageObjects/iosAction.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ Put iOS own actions in this page object.
```
Comment (ellenmobify) [inline app/tests/system/pageObjects/iosAction.js]: `Put iOS-specific actions in this page object.`

File: app/tests/system/pageObjects/iosAction.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	"BACK_BUTTON_XPATH": "//UIAApplication[1]/UIAWindow[1]/UIANavigationBar[1]/UIAButton[2]",
+	"PROFILE_XPATH": "//UIATabBar[1]/UIAButton[4]"
+};
+
+var iOSAction = function(browser) {
+	this.browser = browser;
+	this.selectors = selectors;
+};
+
+iOSAction.prototype.clickBackButton = function(browser) {
```
Comment (ellenmobify) [inline app/tests/system/pageObjects/iosAction.js]: For this and the following functions, you don't need to pass in `browser`.

File: app/tests/system/pageObjects/iosAction.js (hunk 1 above)
Comment (lilyzh) [inline app/tests/system/pageObjects/iosAction.js]: Done.

File: app/tests/system/pageObjects/iosAction.js (hunk 2 above)
Comment (lilyzh) [inline app/tests/system/pageObjects/iosAction.js]: Fixed.

**Comment by ellenmobify** [comment]
Can you fix linting errors on `iosActions.js`?

**Comment by ellenmobify** [comment]
:+1:

**Comment by kschmidtdev** [comment]
@lilyzh I'm still seeing some linting errors:

> tests/system/nightwatch-config.js
>   71:1  warning  Newline required at end of file but not found  eol-last
> tests/system/pageObjects/iosActions.js
>   84:1  warning  Newline required at end of file but not found  eol-last
> tests/system/pages/web/homepage.js
>    3:6   error    Expected indentation of 6 characters           indent
>    6:10  error    Expected indentation of 10 characters          indent
>   11:6   error    Expected indentation of 6 characters           indent
>   13:10  error    Expected indentation of 10 characters          indent
>   16:14  error    Expected indentation of 14 characters          indent
>   22:1   warning  Newline required at end of file but not found  eol-last
> ✖ 8 problems (5 errors, 3 warnings)
> [33mWarning: Task "eslint:dev" failed. Use --force to continue.[39m
> [31mAborted due to warnings.[39m

**Comment by lilyzh** [comment]
@kschmidtdev I just fixed the linting errors. Could you take a look?

**Comment by kschmidtdev** [comment]
The linting issues are fixed, now if I follow the test-drive instructions I get this:

> Error retrieving a new session from the selenium server
> { status: 33,
>   value: 
>    { message: 'A new session could not be created. (Original error: Could not find a device to launch. You requested \'iPhone 6 (9.2 Simulator)\', but the available devices were: ["iPad 2 (9.2) [5CB22C07-FC13-41E2-B1B8-6F6E3C4C37E7]","iPad Air (9.2) [B7C96F3C-BAB2-4817-9556-FA734177B5AF]","iPad Air 2 (9.2) [E3380E7F-5D7A-4261-8C60-9F9C0F3485EA]","iPad Pro (9.2) [37D2A00C-0461-46FC-84D5-FB7CD0B13045]","iPad Retina (9.2) [67875FEB-7D9A-41C8-940B-1D8249B6D761]","iPhone 4s (9.2) [F30656CE-4DEC-4D6D-8B43-12F13FDFC63B]","iPhone 5 (9.2) [57CCBB78-E242-463F-9047-9DFE6A2CB63A]","iPhone 5s (9.2) [EE58F523-6A34-4C92-89EE-0C98AD0B93E0]","iPhone 6 (9.2) [08AF9BEE-5954-4F8F-A235-2DCAA9D9A2D5]","iPhone 6 Plus (9.2) [7E5701BD-9265-4ADC-B5A9-4886D92A9068]","iPhone 6s (9.2) [134E24EB-67A0-4EC1-88E4-80ADF71F856A]","iPhone 6s (9.2) + Apple Watch - 38mm (2.1) [F222561E-FF3F-46A9-86B6-2D8B630AC964]","iPhone 6s Plus (9.2) [DCB78EDF-E6D4-4EDB-9F59-D65E78CBDC5E]","iPhone 6s Plus (9.2) + Apple Watch - 42mm (2.1) [D60C2026-AC8F-4B90-9453-DBBBD0ADEFE2]"])',
>      origValue: 'Could not find a device to launch. You requested \'iPhone 6 (9.2 Simulator)\', but the available devices were: ["iPad 2 (9.2) [5CB22C07-FC13-41E2-B1B8-6F6E3C4C37E7]","iPad Air (9.2) [B7C96F3C-BAB2-4817-9556-FA734177B5AF]","iPad Air 2 (9.2) [E3380E7F-5D7A-4261-8C60-9F9C0F3485EA]","iPad Pro (9.2) [37D2A00C-0461-46FC-84D5-FB7CD0B13045]","iPad Retina (9.2) [67875FEB-7D9A-41C8-940B-1D8249B6D761]","iPhone 4s (9.2) [F30656CE-4DEC-4D6D-8B43-12F13FDFC63B]","iPhone 5 (9.2) [57CCBB78-E242-463F-9047-9DFE6A2CB63A]","iPhone 5s (9.2) [EE58F523-6A34-4C92-89EE-0C98AD0B93E0]","iPhone 6 (9.2) [08AF9BEE-5954-4F8F-A235-2DCAA9D9A2D5]","iPhone 6 Plus (9.2) [7E5701BD-9265-4ADC-B5A9-4886D92A9068]","iPhone 6s (9.2) [134E24EB-67A0-4EC1-88E4-80ADF71F856A]","iPhone 6s (9.2) + Apple Watch - 38mm (2.1) [F222561E-FF3F-46A9-86B6-2D8B630AC964]","iPhone 6s Plus (9.2) [DCB78EDF-E6D4-4EDB-9F59-D65E78CBDC5E]","iPhone 6s Plus (9.2) + Apple Watch - 42mm (2.1) [D60C2026-AC8F-4B90-9453-DBBBD0ADEFE2]"]' },
>   sessionId: null }
> 
> 
> Connection refused! Is selenium server started?
> 
> Fatal error: Tests failed

Maybe I'm missing a simple step?

**Comment by ellenmobify** [comment]
I've seen that issue before on I think certain versions of Appium. Even though you have simulators installed, Appium can't seem to find one to launch. 

I am using Appium v1.4.16 which seems to fix this. Can you try upgrading? If that doesn't work there's a longer workaround..

**Comment by kschmidtdev** [comment]
Thanks @ellenmobify, upgrading fixed my problem!

:+1:

**Pull request merged**